src/tests/ftest/launch.py

```diff
@@ -20,164 +20,552 @@
+    path = os.environ.get("PATH")
+    os.environ["PATH"] = ":".join([bin_dir, sbin_dir, path])
+    os.environ["DAOS_SINGLETON_CLI"] = "1"
+    os.environ["CRT_CTX_SHARE_ADDR"] = "1"
+    os.environ["CRT_ATTACH_INFO_PATH"] = build_vars["PREFIX"] + "/tmp"
+
+    # Verify the PYTHONPATH env is set
+    python_path = os.environ.get("PYTHONPATH")
+    required_python_paths = [
+        os.path.abspath("util/apricot"),
+        os.path.abspath("util"),
+        os.path.abspath("../../utils/py"),
+    ]
+    if python_path is None or python_path == "":
+        os.environ["PYTHONPATH"] = ":".join(required_python_paths)
+    else:
+        defined_python_paths = [
+            os.path.abspath(os.path.expanduser(path))
+            for path in python_path.split(":")]
+        for required_path in required_python_paths:
+            if required_path not in defined_python_paths:
+                python_path += ":" + required_path
+        os.environ["PYTHONPATH"] = python_path
 
-    local_test_files = []
-    test_pattern = "*.py"
 
-    for path, _dirs, files in os.walk(directory):
-        if not (path == directory or path.startswith(os.path.join(directory,
-                                                                  'util'))):
-            for test_file in files:
-                if fnmatch.fnmatch(test_file, test_pattern):
-                    local_test_files.append(os.path.join(path, test_file))
-    return local_test_files
+def get_output(cmd):
+    """Get the output of given command executed on this host.
+
+    Args:
+        cmd (str): command from which to obtain the output
+
+    Returns:
+        str: command output
 
-def yamlforpy(path):
     """
-    Create the name of the yaml file for a given test file.
+    try:
+        print("Running {}".format(cmd))
+        return subprocess.check_output(
+            cmd, stderr=subprocess.STDOUT, shell=True)
+
+    except subprocess.CalledProcessError as err:
+        print("Error executing '{}':\n\t{}".format(cmd, err))
+        exit(1)
+
+
+def time_command(cmd):
+    """Execute the command on this host and display its duration.
+
+    Args:
+        cmd (str): command to time
+
+    Returns:
+        int: return code of the command
+
     """
-    (base, _ext) = os.path.splitext(path)
-    return base + ".yaml"
+    print("Running {}".format(cmd))
+    start_time = int(time.time())
+    return_code = subprocess.call(cmd, shell=True)
+    end_time = int(time.time())
+    print("Total test time: {}s".format(end_time - start_time))
+    return return_code
+
+
+def spawn_commands(host_list, command, timeout=120):
+    """Run the command on each specified host in parallel.
+
+    Args:
+        host_list (list): list of hosts
+        command (str): command to run on each host
+        timeout (int): number of seconds to wait for all jobs to complete
 
-def printhelp():
     """
-    Print a list of test categories.
+    # Create a job to run the command on each host
+    jobs = []
```

Comment:
Indeed.  This work was complete before the move to Clustershell really started and it fixes a pretty serious space usage issue on our hardware clusters that will cause tests to fail eventually so I didn't want to hold it up to re-worked for that.

But indeed, this could/should move to Clustershell.